snakemake/benchmark.py

```diff
@@ -386,3 +388,63 @@ def write_benchmark_records(records, path):
     """Write benchmark records to file at path"""
     with open(path, "wt") as f:
         print_benchmark_records(records, f)
+
+
+def gather_benchmark_records(benchmark_jobs, persistence):
+    """
+    Gather benchmark from given files.
+
+    Args:
+        benchmark_jobs (List[Job]): A list of jobs to extract benchmark from
+    Return:
+        (DataFrame): A Pandas DataFrame object with jobid as index,
+            containing benchmark values for all jobs
+    """
+    benchmarks = pd.DataFrame()
+    for job in benchmark_jobs:
+        if not job._benchmark.exists:
+            raise IOFileException(
+                "Error: \n"
+                + f"Benchmark file {job._benchmark} does not "
+                + f"exist for rule {job.rule.name} \n"
+                + "A complete run is required to print all benchmark metrics."
```

Comment:
Why would we require all jobs to complete? Can we not just leave the output empty for rules that failed, or include an extra column that states "failed/succeeded"? It would be good to know benchmark results for failed jobs as well, as they might have failed due to OOM or exceeding cluster QOS time. This would also serve as a quick way to grep all failed jobs.

Reply:
I agree. But currently Snakemake does not write a benchmark file if a job fails and therefore I think it's a little out of scope for this PR to change that behavior.